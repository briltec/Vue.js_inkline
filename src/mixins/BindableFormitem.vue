<script>
/**
 * Find a form group parent for the given form item
 */
const model = {
    get: function () {
        return this.isGrouped ? this.parentGroup.value : this.value;
    },
    set: function (value) {
        if (this.isGrouped) {
            return this.parentGroup.$emit('change:input', value);
        }

        return this.$emit('input', value);
    }
};

function inputValue () {
    return this.isGrouped ? this.value : this.$attrs.value;
}

export default {
    computed: {
        inputValue,
        model
    }
};
</script>
