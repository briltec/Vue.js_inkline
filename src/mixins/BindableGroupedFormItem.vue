<script>
/**
 * Bind the value of a form item group or groupable form item
 */
const model = {
    get: function () {
        return this.isGrouped ? this.parentGroup.value : this.value;
    },
    set: function (value) {
        if (this.isGrouped) {
            return this.parentGroup.$emit('input', value);
        }

        return this.$emit('input', value);
    }
};

/**
 * Get the value of the form item group or, if it is not grouped, get the value from the individual form item
 */
function currentValue () {
    return this.isGrouped ? this.value : this.$attrs.value;
}

export default {
    computed: {
        currentValue,
        model
    }
};
</script>
