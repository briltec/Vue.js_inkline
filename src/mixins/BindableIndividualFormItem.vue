<script>
/**
 * Bind the value of an individual form item
 */
const model = {
    get: function () {
        return this.value;
    },
    set: function (value) {
        return this.$emit('input', value);
    }
};

/**
 * Get the value of the individual form item
 */
function currentValue () {
    return this.value;
}

export default {
    computed: {
        currentValue,
        model
    }
};
</script>
