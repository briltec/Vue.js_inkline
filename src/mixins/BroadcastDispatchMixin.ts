/* eslint-disable no-invalid-this, @typescript-eslint/ban-ts-comment */
import { defineComponent } from 'vue';

/**
 * Broadcast an event with given params to specific child properties
 *
 * @param componentName
 * @param eventName
 * @param params
 */
export function broadcast (this: any, componentName: string, eventName: string, params: any) {
    this.$children.forEach((child: any) => {
        const name = child.$options.name;

        if (name === componentName) {
            child.$emit.apply(child, [eventName].concat(params));
        } else {
            // @ts-ignore
            broadcast.apply(child, [componentName, eventName].concat([params]));
        }
    });
}

/**
 * Dispatch an event from child to parents of given type
 *
 * @param componentName
 * @param eventName
 * @param params
 */
export function dispatch (componentName: string, eventName: string, params: any) {
    // @ts-ignore
    let parent = this.$parent || this.$root;
    let name = parent.$options.name;

    while (parent && (!name || name !== componentName)) {
        parent = parent.$parent;

        if (parent) {
            name = parent.$options.name;
        }
    }
    if (parent) {
        parent.$emit.apply(parent, [eventName].concat(params));
    }
}

export default defineComponent({
    methods: {
        dispatch (componentName: string, eventName: string, params: any) {
            dispatch.call(this, componentName, eventName, params);
        },
        broadcast (componentName: string, eventName: string, params: any) {
            broadcast.call(this, componentName, eventName, params);
        }
    }
});
