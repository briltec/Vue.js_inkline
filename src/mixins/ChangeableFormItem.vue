<script>
/**
 * Handler for change event
 */
function onChange (event) {
    this.$emit('change', event.target.value);
}
export default {
    methods: {
        onChange
    }
};
</script>
