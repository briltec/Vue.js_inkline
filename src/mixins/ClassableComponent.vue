<script>
export default {
    data () {
        return {
            classesProvider: {
                root: [],
                child: []
            }
        };
    },
    computed: {
        /**
         * Compute dynamically provided classes from mixins
         */
        classes () {
            const classRules = {
                '*': []
            };

            Object.keys(this.classesProvider).forEach((key) => {
                classRules[key] = this.classesProvider[key].map((classRules) => classRules());
                classRules['*'] = classRules['*'].concat(classRules[key]);
            });

            return classRules;
        }
    }
};
</script>
