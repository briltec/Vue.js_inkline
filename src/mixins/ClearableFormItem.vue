<script>
export default {
    props: {
        clearable: {
            type: Boolean,
            default: false
        }
    },
    computed: {
        isClearable () {
            return this.clearable && this.currentValue !== '' && (this.focused || this.hovered);
        }
    },
    methods: {
        clear () {
            this.model = '';
            this.$emit('clear');
            this.focus();
        }
    },
    created () {
        if (this.classesProvider) {
            this.classesProvider.push(() => ({
                '-clearable': this.custom
            }));
        }
    }
};
</script>
