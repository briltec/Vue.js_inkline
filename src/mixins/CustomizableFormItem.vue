<script>
export default {
    created () {
        this.classRules.push({ '-custom': this.custom });
    },
    props: {
        custom: {
            type: Boolean,
            default: true
        }
    }
};
</script>
