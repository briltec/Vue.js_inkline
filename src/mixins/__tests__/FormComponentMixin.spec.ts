import { FormComponentMixin } from '@inkline/inkline/mixins';
import { render } from '@testing-library/vue';
import { createMockInstance } from '@inkline/inkline/__mocks__/createMockInstance';

describe('mixins', () => {
    describe('FormComponentMixin', () => {
        const Component = {
            name: 'FormComponent',
            mixins: [FormComponentMixin],
            template: '<input />'
        };

        it('should render correctly', () => {
            const wrapper = render(Component);

            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('computed', () => {
            describe('isDisabled', () => {
                it('should be true if disabled', () => {
                    const wrapper = createMockInstance(FormComponentMixin);
                    wrapper.disabled = true;

                    expect(wrapper.isDisabled).toEqual(true);
                });

                it('should be true if form.isDisabled', () => {
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            form: {
                                isDisabled: true
                            }
                        }
                    });

                    expect(wrapper.isDisabled).toEqual(true);
                });

                it('should be true if formGroup.isDisabled', () => {
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup: {
                                isDisabled: true
                            }
                        }
                    });

                    expect(wrapper.isDisabled).toEqual(true);
                });
            });

            describe('isReadonly', () => {
                it('should be true if disabled', () => {
                    const wrapper = createMockInstance(FormComponentMixin);
                    wrapper.readonly = true;

                    expect(wrapper.isReadonly).toEqual(true);
                });

                it('should be true if form.isReadonly', () => {
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            form: {
                                isReadonly: true
                            }
                        }
                    });

                    expect(wrapper.isReadonly).toEqual(true);
                });

                it('should be true if formGroup.isReadonly', () => {
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup: {
                                isReadonly: true
                            }
                        }
                    });

                    expect(wrapper.isReadonly).toEqual(true);
                });
            });

            describe('parent', () => {
                it('should be formGroup if formGroup instantiated', () => {
                    const formGroup = {
                        $: true
                    };
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup
                        }
                    });

                    expect(wrapper.parent).toEqual(formGroup);
                });

                it('should be form if formGroup not instantiated', () => {
                    const form = {};
                    const formGroup = {
                        $: false
                    };
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup,
                            form
                        }
                    });

                    expect(wrapper.parent).toEqual(form);
                });
            });

            describe('schema', () => {
                it('should be parent schema if name not provided', () => {
                    const formGroup = {
                        $: true,
                        schema: {}
                    };
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup
                        }
                    });
                    wrapper.name = '';

                    expect(wrapper.schema).toEqual(formGroup.schema);
                });

                it('should be default schema if name and parent schema not provided', () => {
                    const formGroup = {
                        $: true
                    };
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup
                        }
                    });
                    wrapper.name = '';

                    expect(wrapper.schema).toEqual({});
                });

                it('should get schema using name path if name provided', () => {
                    const formGroup = {
                        $: true,
                        schema: {
                            field: {}
                        }
                    };
                    const wrapper = createMockInstance(FormComponentMixin, {
                        inject: {
                            formGroup
                        }
                    });
                    wrapper.name = 'field';

                    expect(wrapper.schema).toEqual(formGroup.schema.field);
                });
            });
        });
    });
});
