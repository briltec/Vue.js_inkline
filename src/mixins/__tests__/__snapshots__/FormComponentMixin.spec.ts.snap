// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mixins FormComponentMixin should render correctly 1`] = `<input>`;
